src/test/java/uk/gov/ons/ctp/integration/envoycuc/mockclient/MockClient.java

```
@@ -0,0 +1,292 @@
+    return response;
+  }
+
+  public IsValidRequest postRequest(final RateLimiterClientRequest rateLimiterClientRequest)
+      throws ResponseStatusException {
+
+    List<String> requestKeyList = getKeys(rateLimiterClientRequest);
+    final IsValidRequest isValidRequest =
+        isValidateRequest(requestKeyList, rateLimiterClientRequest);
+    postRequest(
+        requestKeyList,
+        rateLimiterClientRequest); // always post - it burns allowances every time for all scenarios
+
+    return isValidRequest;
+  }
+
+  private List<String> getKeys(RateLimiterClientRequest request) {
+    List<String> keyList = new ArrayList<>();
+    StringBuilder keyBuff =
+        new StringBuilder(request.getProduct().getProductGroup().name().toUpperCase())
+            .append("-")
+            .append(request.getProduct().getIndividual().toString().toUpperCase())
+            .append("-")
+            .append(request.getProduct().getDeliveryChannel().name().toUpperCase())
+            .append("-")
+            .append(request.getCaseType().name().toUpperCase())
+            .append("-");
+    if (request.getUprn() != null && request.getUprn().getValue() != 0) {
+      keyList.add(keyBuff.toString() + "UPRN");
+    }
+    if (request.getIpAddress() != null) {
+      keyList.add(keyBuff.toString() + "IP");
+    }
+    if (request.getTelNo() != null) {
+      keyList.add(keyBuff.toString() + "TELNO");
+    }
+    return keyList;
+  }
+
+  private IsValidRequest isValidateRequest(
+      List<String> requestKeyList, RateLimiterClientRequest f) {
+    final IsValidRequest isValidRequest = new IsValidRequest();
+    for (String requestKey : requestKeyList) {
+      if (!allowanceMap.containsKey(requestKey)) {
+        continue; // not in scope so is valid....
+      }
+      final int noRequestAllowed = allowanceMap.get(requestKey);
+      Map<String, List<Integer>> postedMap = postingsTimeMap.get(requestKey);
+      if (postedMap == null) {
+        return isValidRequest;
```

do we want to throw an exception here? since we have slipped through  finding an allowance, so if we can't find a tracking map , I imagine this is an error in the code?

---

The question is - what does the real limiter do?